sdk/go/common/workspace/plugins_install_test.go

```diff
@@ -58,48 +61,163 @@ func createTGZ(files map[string][]byte) ([]byte, error) {
+	defer os.RemoveAll(dir)
 
-	finalDirRoot, err := ioutil.TempDir("", "final-dir")
+	err := plugin.Install(tarball)
 	assert.NoError(t, err)
-	defer os.RemoveAll(finalDirRoot)
-	finalDir := filepath.Join(finalDirRoot, "final")
 
-	err = installPlugin(finalDir, ioutil.NopCloser(bytes.NewReader(tgz)))
-	assert.NoError(t, err)
+	assertPluginInstalled(t, dir, plugin)
 
-	info, err := os.Stat(filepath.Join(finalDir, expectedDir))
+	info, err := os.Stat(filepath.Join(dir, plugin.Dir(), expectedDir))
 	assert.NoError(t, err)
 	assert.True(t, info.IsDir())
+
+	testDeletePlugin(t, dir, plugin)
 }
 
 func TestInstallNoDeps(t *testing.T) {
 	name := "foo.txt"
 	content := []byte("hello\n")
 
-	tgz, err := createTGZ(map[string][]byte{name: content})
+	dir, tarball, plugin := prepareTestDir(t, map[string][]byte{name: content})
+	defer os.RemoveAll(dir)
+
+	err := plugin.Install(tarball)
 	assert.NoError(t, err)
 
-	finalDirRoot, err := ioutil.TempDir("", "final-dir")
+	assertPluginInstalled(t, dir, plugin)
+
+	b, err := ioutil.ReadFile(filepath.Join(dir, plugin.Dir(), name))
 	assert.NoError(t, err)
-	defer os.RemoveAll(finalDirRoot)
-	finalDir := filepath.Join(finalDirRoot, "final")
+	assert.Equal(t, content, b)
+
+	testDeletePlugin(t, dir, plugin)
+}
+
+func TestInstallCleansOldFiles(t *testing.T) {
+	dir, tarball, plugin := prepareTestDir(t, nil)
+	defer os.RemoveAll(dir)
 
-	err = installPlugin(finalDir, ioutil.NopCloser(bytes.NewReader(tgz)))
+	// Leftover temp dirs.
+	tempDir1, err := ioutil.TempDir(dir, fmt.Sprintf("%s.tmp", plugin.Dir()))
+	assert.NoError(t, err)
+	tempDir2, err := ioutil.TempDir(dir, fmt.Sprintf("%s.tmp", plugin.Dir()))
+	assert.NoError(t, err)
+	tempDir3, err := ioutil.TempDir(dir, fmt.Sprintf("%s.tmp", plugin.Dir()))
 	assert.NoError(t, err)
 
-	info, err := os.Stat(filepath.Join(finalDir, name))
+	// Leftover partial file.
+	partialPath := filepath.Join(dir, plugin.Dir()+".partial")
+	err = ioutil.WriteFile(partialPath, nil, 0600)
 	assert.NoError(t, err)
-	assert.False(t, info.IsDir())
 
-	b, err := ioutil.ReadFile(filepath.Join(finalDir, name))
+	err = plugin.Install(tarball)
 	assert.NoError(t, err)
-	assert.Equal(t, content, b)
+
+	assertPluginInstalled(t, dir, plugin)
+
+	// Verify leftover files were removed.
+	for _, path := range []string{tempDir1, tempDir2, tempDir3, partialPath} {
+		_, err := os.Stat(path)
+		assert.Error(t, err)
+		assert.True(t, os.IsNotExist(err))
+	}
+
+	testDeletePlugin(t, dir, plugin)
+}
+
+func TestGetPluginsSkipsPartial(t *testing.T) {
+	dir, tarball, plugin := prepareTestDir(t, nil)
+	defer os.RemoveAll(dir)
+
+	err := plugin.Install(tarball)
+	assert.NoError(t, err)
+
+	err = ioutil.WriteFile(filepath.Join(dir, plugin.Dir()+".partial"), nil, 0600)
+	assert.NoError(t, err)
+
+	assert.False(t, HasPlugin(plugin))
+
+	has, err := HasPluginGTE(plugin)
+	assert.Error(t, err)
+	assert.False(t, has)
+
+	plugins, err := getPlugins(dir)
+	assert.Equal(t, 0, len(plugins))
```

Comment:
Consider adding a test that tries to run multiple installations in parallel and validates that the result is consistent.

Reply:
Done.